testing/python/metafunc.py

```diff
@@ -2114,3 +2114,10 @@ def test_converted_to_str(a, b):
                 "*= 6 passed in *",
             ]
         )
+
+    def test_param_usefixtures_rejected(self, pytester: Pytester) -> None:
+        """Verify that passing usefixtures to pytest.param fails loudly rather than silently."""
+        pytest.raises(
+            ValueError,
+            lambda: pytest.param("x", [1], marks=pytest.mark.usefixtures("foo")),
+        )
```

Comment:
I think it would be better for us to have an integration test instead, because in this it is important for us to see how the error will appear to the user: for example, it should not blow up an entire traceback with the pytest internals, it should show a succinct message pointing out the fixture, location, etc.

I would start by using `pytest.fail(..., pytrace=False)` or raising `UsageError`, whichever produces a better result.

Reply:
@nicoddemus `UsageError` and `ValueError` look the same to me, like this (other than the exception name, obviously):

```
___________________ ERROR collecting test_foo.py _____________________
test_foo.py:6: in <module>
    pytest.param(
/usr/lib/python3.12/site-packages/_pytest/mark/__init__.py:72: in param
    return ParameterSet.param(*values, marks=marks, id=id)
/usr/lib/python3.12/site-packages/_pytest/mark/structures.py:92: in param
    raise ValueError(
E   ValueError: pytest.param cannot add pytest.mark.usefixtures; see https://github.com/pytest-dev/pytest/issues/4112
```

while `pytest.fail` looks like this:

```
___________________ ERROR collecting test_foo.py _____________________
pytest.param cannot add pytest.mark.usefixtures; see https://github.com/pytest-dev/pytest/issues/4112
```

So it looks like `pytest.fail` produces the least output (and without any Pytest internals mentioned). However, it also doesn’t mention the line number on the user test file where the error occurred (remember this happens at import time, not execution time, so it also doesn’t mention the function name), which the others (since they include traces) do. Is that a problem? It seems to me it would be useful to point the user at the specific `param` call that has the problem, but if you prefer the shorter output, I’ll go ahead with that.

So, given the above tradeoff, what’s your preference? `ValueError`, `UsageError`, or `pytest.fail`?

Reply:
Hmm I see thanks.

Let's keep it `ValueError` for now then, I agree that the location/line number is really important -- due to how `pytest.param` is exposed in the API, this is reasonable to do for now (we could do some `sys.frame()` shenaningans to get the `pytest.param` location invocation and build up a message ourselves, but not sure it is worth doing that for this quite uncommon occurrence).

But we should move forward with the integration test I mentioned: it is important that we check that `test_foo.py:6` and `pytest.param(` appear in the message (from your example posted above).

Also, we probably should put a link to the docs somewhere explaining that (possibly we can add a note to `pytest.param` documentation), rather than pointing to the GitHub issue.

Reply:
Yep, I will definitely do the integration test change, just waiting to do that until we’ve nailed down what we want the behaviour to be (no point changing the test twice).

In terms of a link to documentation, my understanding was that one *should* be able to `usefixtures` in `param`, but it just didn’t work right now due to a bug. Is it actually intended that `usefixtures` not be used in `param`? Or do you think, even though it’s a bug, it should be in the manual (typically I don’t see bugs documented in manuals, but rather only bug trackers, on most projects)?

Reply:
> Yep, I will definitely do the integration test change, just waiting to do that until we’ve nailed down what we want the behaviour to be (no point changing the test twice).

Oh sure, thanks, just wanted to take the opportunity to point out what I would like to see checked in the integration test.

> Is it actually intended that usefixtures not be used in param?

We don't have any plans to support `usefixtures` there (nor in fixtures), so we might as well place the error (as you are doing) and leave a note in the docs (even if that changes in the future, I think it is harmless to let users know the current situation explicitly).

Reply:
I was hoping to make fixing this one commit, but I had no idea CI was going to come in and add extra commits, so that didn’t work out. So this is now b3f232a800108c32368c030dcd7b39990240188a and ccc1e18702117d5671f5553a420eadef68405edb, plus anything else that needs to follow it.